src/doorman/Fd.cpp

```diff
@@ -0,0 +1,86 @@
+#include "Fd.hpp"
+
+#include <fmt/format.h>
+
+using namespace fmt::literals;
+
+void Fd::write(const void *data, size_t length) const {
+    if (!is_open())
+        throw std::runtime_error("Write called on invalid file descriptor");
+    auto num = ::write(fd_, data, length);
+    if (num < 0)
+        throw fmt::system_error(errno, "Unable to write to {}", fd_);
+    if (static_cast<size_t>(num) != length)
+        throw std::runtime_error("Truncated write to file descriptor {}"_format(fd_));
+}
+
+void Fd::writev(gsl::span<const iovec> io_vecs) const {
+    if (!is_open())
+        throw std::runtime_error("Write called on invalid file descriptor");
+    auto num = ::writev(fd_, io_vecs.data(), io_vecs.size());
+    if (num < 0)
+        throw fmt::system_error(errno, "Unable to write to {}", fd_);
+
+    size_t length = 0;
+    for (auto &v : io_vecs)
+        length += v.iov_len;
+
+    if (static_cast<size_t>(num) != length)
+        throw std::runtime_error("Truncated write to file descriptor {}"_format(fd_));
+}
+
+Fd Fd::accept(sockaddr *address, socklen_t *socklen) const {
+    auto accepted_fd = ::accept(fd_, address, socklen);
+    if (accepted_fd < 0)
+        throw fmt::system_error(errno, "Unable to accept from file descriptor {}", fd_);
+    return Fd(accepted_fd);
+}
+
+Fd Fd::socket(int domain, int type, int protocol) {
+    int socket_fd = ::socket(domain, type, protocol);
+    if (socket_fd < 0)
+        throw fmt::system_error(errno, "Unable to create a socket");
```

Comment:
Include domain, type, protocol in the message?

Reply:
Done, only as numeric though so useful but not as useful as the name (again no code I know of exists to turn 4 into "AF_INET" but it's better than nothing).